rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,510 @@
+    </ComponentCustomStylesContextProvider>
+  </FluentProvider>
+</App>
+```
+
+### 👍
+
+- Custom styling gets the same power as the default component useStyles hook.
+- This builds on using slots as the addressable items for applying styles.
+- The custom styles can be built using Griffel with build-time optimization.
+- Can be applied at different scope like FluentProvider.
+- Does not require composing or introducing a new component type that must be adopted in lieu of the default component.
+- Fewer CSS specificity problems as styles can get overwritten.
+- Provides the type-safety expected from Fluent UI React.
+- Minimal code to add to each component. Could be put into the code generation template.
+- Does not significantly affect rendering performance.
+- Partners building components can follow the same practice to support custom styling in their libraries.
+
+### 👎
+
+- Custom styling all components requires adding and configuring lots of context providers.
+- Requires writing new code for every component, adding a story, and documenting the feature.
+- Context is harder to discover as it is outside the props contract.
+- Callers doing custom styling may still take a dependency on the CSS internals of the component if they don't replace a
+  style block completely.
+
+### 🤔 Centralized and singular custom styling method
+
+It would be nice to have a single method on context that could apply for all components. This proves difficult
+with each component having its own props and state types.
+
+Another approach would be to define single context value type in the react-shared-contexts package that had
+a method for each component, but that causes a circular dependency problem for components.
+
+There may some typescript trickery that would allow for a single context value type, but having each component define
+the context type it will inspect for custom styling follows SOLID dependency inversion.
+
+## Option A2: Add global custom styling hook
+
+A global custom styling hook would be defined. It takes a component kind and state.
+
+> We experimented with several ways to try and make the method typesafe, but it proved
+> either overly complex, forced conditional calls that would make hooks unstable,
+> led to cirtcular references between packages, or restricted custom styling to the point
+> of lower value as an escape hatch.
+
+```tsx
+export type FluentStyleCustomizer = {
+  useCustomStyles_unstable: (kind: string, state: any) => void;
+};
+
+export const fuiCustomizer: FuiCustomizer = {
+  useCustomStyles_unstable: () => {},
+};
+```
+
+Components would call the custom styling hook immediately after their own styling hook.
+
+> The kind passed to the custom style hook can be specific to library (e.g. 'FluentUI.Button').
+
+```tsx
+export const SomeComponent: ForwardRefComponent<SomeComponentProps> = React.forwardRef((props, ref) => {
+  const state = useSomeComponent_unstable(props, ref);
+
+  useSomeComponentStyles_unstable(state);
```

Comment:
I would like to discuss whether we always call the base styles hook.

Forget the rule of hooks for a second, why don't we just call the custom styles if it is set in the context and let it up to the partner to decide whether they want to call the original function and merge it with overrides or completely replace the styles with the custom ones (which would allow them to also use `useResetStyles()`?

Reply:
My only concern is that conditionally calling the custom style hook may lead people to unstable hook calls.  Of course, if the hook is set at app start then the number of hooks called would be the same across renders, but the conditional call still looks suspicious :)